### PR TITLE
WSGI Middleware on the JVM

### PR DESCRIPTION
**Pull request by dmnjohns**
Author(s) mini bio:
My name is Damien Johnson. I've been a software developer at Rackspace for nearing three years now, after graduating with my bachelor's degree in Computer Science from UT Austin.

Author(s) social media contact info:
www.linkedin.com/in/dmnjohns
https://plus.google.com/u/0/+DamienJohnson

Accept or decline to allow redaction, editing or structure changes by technical writer staff:
Accept

Desired date of publishing:
09/04/2015

### CODE REVIEW

**Comment by smashwilson** [comment]
Er, did this one split through the cracks? Sorry about that!

@catlook: This one can use some red pen too :wink:

**Comment by dmnjohns** [comment]
It must have, but it's not a problem. Honestly, I've been too busy to check back on it!

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+![The unicorn of the sea]({% asset_path repose_logo.png %})
+
+This is the second in a series of posts written by the [Repose][repose] Ninja
+on Duty. Special thanks to [Jim Baker](https://github.com/jimbaker/): author of
+the [Fireside][fireside] project, Jython core contributor, and integral braniac
+in the effort to support the WSGI specification through Servlet technologies.
+
+If you have ever had the pleasure of evaluating [Repose][repose], you may have
+noticed that while it provides an incredibly powerful foundation, it is
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ..  that, while it provides ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+missing that one all-important feature that you need. While the [Repose][repose]
+team does its best to handle all common and reasonable use-cases, there are an
+infinite number of problems for which [Repose][repose] is a solution, therefore
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... a solution. Therefore ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+it is impossible to predict and develop features to solve every problem.
+Luckily, [Repose][repose] is built on a pluggable architecture that any
+developer can leverage to solve the problem of the day.
+
+In this post, I will expand upon the previous post in this series by diving
+deeper into the [Repose][repose] extensibility model, and explaining how
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ...  model and explaining ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+[Repose][repose] plans to make that model more developer friendly in the future.
+
+<!-- more -->
+
+# The Repose Extensibility Model
+It is important to note that Repose was designed with extensibility in mind.
+Built on top of Java's Servlet technology, Repose supports loading and running
+custom Servlet Filters. These filters have full access to the HTTP request and
+response, and may perform any processing that is desired. They are even able to
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... and they may perform ..

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+PowerFilter, in turn, manages all of the filters which compose Repose. What
+this means is that the lifecycle of custom filters are tied to the lifecycle
+of the PowerFilter, but the PowerFilter has additional control over other
+filters. To be more specific, the PowerFilter will destroy all filters and
+create new ones any time the system model configuration is changed, or an
+artifact in the artifact directory is changed. For code that is meant to be
+long-running, this behavior potentially poses an issue. To combat this issue,
+Repose uses services. As you might expect, a service is simply a long-running,
+shared component. At the time of writing this post, Repose does not support
+custom services, but is expected to in the relatively near future.
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... but it is expected to ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+Jython's dynamic language properties and how those properties interplay with
+Java's JIT compilation. I suspect that, in the future, we'll see convergence in
+performance between these three languages.
+
+To learn more about Jython in its current and future state, check out [this
+terrific talk by Jim Baker at PyCon](https://youtu.be/hLm3garVQFo).
+
+# Bridging the Gap between WSGI and Servlet Specifications
+This is where things get interesting. The Servlet spec and WSGI spec were born
+from two very different environments, and as a result, expect very different
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... environments and, as a result, expect very ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+interactions. While I won't go into the history for each, I will give brief
+descriptions.
+
+Let's start with the Servlet spec. The Servlet spec defines a Java component
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... Java component, ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+to run alongside Java code on the JVM. Once we start using Jython, importing
+and using Java Servlet classes is trivial, and calling WSGI-compliant apps is
+even simpler. With the language barrier out of the way, let's look at the spec
+barrier.
+
+To start, since Repose will be passing Servlet objects, we'll have to convert
+those into the format a WSGI app expects. The naive solution would be to copy
+all of the data out of the Servlet objects into a dict which gets passed to a
+WSGI app. Of course, that can get expensive, and there really isn't a need.
+Instead, we can build a lazy, caching map adapter which will take WSGI CGI-like
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... map adapter, which will ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+we're constrained by Repose, we have to do a bit of work. While the bridge
+described above allows our WSGI code to read from Servlet objects, we still
+have to write to them. Normally, in WSGI apps, we simply callback to the
+start_response callable to set the status code and headers, then return an
+iterable which provides our body from the call into our app. Effectively,
+the flow in a WSGI app is a series of composed functions. In Servlet spec
+land, however, we must mutate the response object. We're given some structure
+in the form of the FilterChain doFilter method and the Servlet service method,
+but really, these methods only exist to define an interface. To reconcile these
+two approaches, we must write a bridge which understands both specifications,
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... both specifications ...     [no comma at end]

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+and can translate between the two. The approach that was taken in Fireside
+wraps a WSGI app with a Servlet Filter which will provide a callback to the
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: ... Servlet Filter, which will ...

File: _posts/2015-09-04-wsgi-middleware-on-the-jvm.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+Before I conclude, I have to take this opportunity to recommend that you go
+check out Repose if you haven't yet! It's not a silver bullet, but it can do
+quite a bit, and it's getting better every day. Running WSGI middleware is just
+one of many enhancements Coming Soon&trade;.
+
+If you have any questions, would like to know more, or would like to get
+involved in the this effort, please feel free to
+[send me an email](mailto:damien.johnson@rackspace.com?subject=WSGI%20Middleware%20On%20The%20JVM)
+and/or [check out the Repose project!](https://github.com/rackerlabs/repose)
+
```
Comment (catlook) [inline _posts/2015-09-04-wsgi-middleware-on-the-jvm.md]: [ ed. note:  nice article!  Use edits you like and discard the rest :) ]

**Comment by dmnjohns** [comment]
I've made some revisions in response to @catlook's comments. I've also bumped the publish date. Thanks for the help, and let me know if there's anything else I need to do!

**Comment by smashwilson** [comment]
On its way out :shipit:

**Pull request merged**